src/eventer/eventer_aco_opset.c

```diff
@@ -0,0 +1,303 @@
+  eventer_t ae = eventer_alloc_asynch(eventer_aco_asynch_wrapper, ctx);
+  if(eventer_try_add_asynch_subqueue(q, ae, sq)) {
+    aco_yield();
+    return mtev_true;
+  }
+  free(ctx);
+  return mtev_false;
+}
+
+void
+eventer_aco_run_asynch_queue_subqueue(eventer_jobq_t *q, eventer_t e, uint64_t sq) {
+  struct aco_asynch_cb_ctx *ctx = calloc(1, sizeof(*ctx));
+  ctx->e = e;
+  ctx->co = aco_get_co();
+  eventer_t ae = eventer_alloc_asynch(eventer_aco_asynch_wrapper, ctx);
+  eventer_add_asynch_subqueue(q, ae, sq);
+  aco_yield();
+}
+
+static int
+aco_timeout(eventer_t e, int mask, void *closure, struct timeval *now) {
+  aco_opset_info_t *info = closure;
+  struct aco_cb_ctx *ctx = info->aco_co->arg;
+  ctx->timeout_e = NULL;
+  ctx->private_errno = ETIME;
+  eventer_aco_resume(info->aco_co);
+  return 0;
+}
+
+void
+eventer_aco_sleep(struct timeval *timeout) {
+  mtevAssert(aco_get_co());
+  struct aco_cb_ctx *ctx = aco_get_arg();
+  mtevAssert(!ctx->timeout_e);
+  aco_opset_info_t *info = calloc(1, sizeof(*info));
+  info->aco_co = aco_get_co();
+  ctx->timeout_e = eventer_in(aco_timeout, info, *timeout);
+  eventer_add(ctx->timeout_e);
+  aco_yield();
+  free(info);
+}
+
+#define ACO_TIMEOUT_CALL(name, params, args) \
+static int \
+aco_##name params { \
+  eventer_t e = closure; \
+  aco_opset_info_t *info = (aco_opset_info_t *)e->opset_ctx; \
+  struct aco_cb_ctx *ctx = aco_get_arg(); \
+  int added = 0; \
+\
+  mtevAssert(aco_get_co() == info->aco_co); \
+  mtevAssert(!ctx->timeout_e); \
+  if(ctx->timeout) { \
+    ctx->timeout_e = eventer_in(aco_timeout, info, *ctx->timeout); \
+    ctx->timeout = NULL; \
+    eventer_add(ctx->timeout_e); \
+  } \
+\
+  while(1) { \
+    if(added) { eventer_remove(e); added = 0; } \
+    ctx->rv = info->original_opset->name args; \
+    if(ctx->rv == -1 && (errno == EINPROGRESS || errno == EAGAIN)) { \
+      e->mask = *mask; \
+      ctx->mask = *mask; \
+      if(!added) { eventer_add(e); added = 1; } \
+      aco_yield(); \
+      if(ctx->private_errno == ETIME) { \
+        errno = ctx->private_errno; \
+        ctx->private_errno = 0; \
+        ctx->rv = -1; \
+        break; \
+      } \
+    } \
+    else break; \
+  } \
+  if(ctx->timeout_e) { \
+    mtevEvalAssert(eventer_remove(ctx->timeout_e)); \
+    eventer_free(ctx->timeout_e); \
+    ctx->timeout_e = NULL; \
+  } \
+  if(added) { \
+    mtevEvalAssert(eventer_remove(e)); \
+  } \
+  return ctx->rv; \
+}
+
+ACO_TIMEOUT_CALL(accept,
+                 (int fd, struct sockaddr *addr, socklen_t *len, int *mask, void *closure),
+                 (fd, addr, len, mask, closure));
+ACO_TIMEOUT_CALL(read,
+                 (int fd, void *buffer, size_t len, int *mask, void *closure),
+                 (fd, buffer, len, mask, closure));
+ACO_TIMEOUT_CALL(write,
+                 (int fd, const void *buffer, size_t len, int *mask, void *closure),
+                 (fd, buffer, len, mask, closure));
+
+static int
+aco_close(int fd,
```

Comment:
I am worried these might in the future collide with the `aco` namespace (`aco.h`).. maybe prefix these functions with `eventer_`?